lib/portage/dep/libc.py

```diff
@@ -0,0 +1,70 @@
+# Copyright 2023 Gentoo Authors
+# Distributed under the terms of the GNU General Public License v2
+
+from portage.dep import Atom
+from portage.const import LIBC_PACKAGE_ATOM
+from portage.dbapi._expand_new_virt import expand_new_virt
+
+import portage.dbapi.porttree
+
+
+def find_libc_deps(portdb: portage.dbapi.porttree.dbapi, realized: bool = False):
+    """Finds libc package for a ROOT via portdb.
+
+    Parameters
+    ----------
+    portdb : dbapi
+        dbapi instance for portdb (for installed packages).
+    realized : bool
+        Request installed atoms rather than the installed package satisfying LIBC_PACKAGE_ATOM.
+
+    Returns
+    -------
+    list
+        List of libc packages (or atoms if realized is passed).
+    """
+
+    libc_pkgs = set()
+
+    for atom in expand_new_virt(
+        portdb,
+        LIBC_PACKAGE_ATOM,
+    ):
+        if atom.blocker:
+            continue
+
+        if not realized:
+            # Just the raw packages were requested (whatever satifies the virtual)
+            libc_pkgs.add(atom)
+            continue
+
+        # This will give us something like sys-libs/glibc:2.2, but we want to know
+        # what installed atom actually satifies that.
+        try:
+            libc_pkgs.add(portdb.match(atom)[0])
+        except IndexError:
+            continue
+
+    return libc_pkgs
+
+
+def strip_libc_deps(dep_struct, libc_deps):
+    """TODO"""
+    # We're going to just grab the libc provider for ROOT and
+    # strip out any dep for the purposes of --changed-deps.
+    # We can't go off versions, even though it'd be more precise
+    # (see below), because we'd end up with FPs and unnecessary
+    # --changed-deps results far too often.
+    #
+    # This penalizes a bit the case where someone adds a
+    # minimum (or maximum) version of libc explicitly in an ebuild
+    # without a new revision, but that's extremely rare, and doesn't
+    # feel like it changes the balance for what we prefer here.
+    for i, x in enumerate(dep_struct):
+        # We only need to bother if x is an Atom because we know the deps
+        # we inject are simple & flat.
+        if isinstance(x, Atom) and any(
+            libc_dep.without_slot.match(x) for libc_dep in libc_deps
```

Comment:
I've just realized that it's only by luck that this hasn't triggered an exception due to some assumption in match_from_list, since match_from_list is intended to match an atom against packages, not atoms against atoms.  For example, if you look through match_from_list you'll see that it assumes packages have a cpv_split attribute which atoms do not have. The function originally operated on simple "cpv" packages, and later gained support for Package and _pkg_str instances.

Reply:
The Atom.intersects method is the closest we have to what you want here.

Reply:
I'm wondering what to do about this. The intention is to only call `strip_libc_deps` with "realized" dependencies, which should be simple atoms.

So, the current approach is okay only if `find_libc_deps` can never return a complex dependency with an operator like '=*'.

Can we rely on `expand_new_virt` with portdb to only return things we can use as an atom, or no?

Reply:
The intention is to only call strip_libc_deps with "realized" dependencies, which should be simple atoms.

You say "realized" but strip_libc_deps is currently called with realized=False just prior to strip_libc_deps.

`expand_new_virt` with portdb will always return a sequence of valid atoms, which may include atoms with any operators including '=*'.

Reply:
@zmedico do you have a suggestion on how do do this correctly?

Reply:
I think we can do it with the `intersects` method instead of the `match` method , at least until someone comes up with a better idea.

Reply:
Thanks. I have a bunch of local changes I've been chipping away at but gone around in circles a bit. I'll try tidy them up over the weekend.

I also need to see if we have a test I can borrow for merging a binpkg so I can inspect its metadata to check the whole thing works OK.

Reply:
I think we're good here now, fingers crossed.